In [277]:
# Homework 2. Daniil Morozkov. 09/05/2023.
# import libraries
import pandas as pd
import numpy as np

Question 1
Using the full SPY dataset, compute daily close-to-close returns on SPY.  What is the average return on SPY over the full sample, during the first trading day of the month, and during the third-to-the-last trading day of the month?


In [278]:
# upload dataset, show few entries
df_spy = pd.read_csv('SPY.csv', parse_dates=[0], index_col=[0])
df_spy.head()

Open      High       Low     Close  Adj Close   Volume
Date                                                                  
1993-01-29  43.96875  43.96875  43.75000  43.93750  25.438095  1003200
1993-02-01  43.96875  44.25000  43.96875  44.25000  25.619024   480500
1993-02-02  44.21875  44.37500  44.12500  44.34375  25.673296   201300
1993-02-03  44.40625  44.84375  44.37500  44.81250  25.944683   529400
1993-02-04  44.96875  45.09375  44.46875  45.00000  26.053244   531500

In [279]:
# find returns
df_spy['Return'] = df_spy['Adj Close'].pct_change()
df_spy['Return']

Date
1993-01-29         NaN
1993-02-01    0.007113
1993-02-02    0.002118
1993-02-03    0.010571
1993-02-04    0.004184
                ...   
2022-07-20    0.006373
2022-07-21    0.010183
2022-07-22   -0.009278
2022-07-25    0.001215
2022-07-26   -0.011831
Name: Return, Length: 7426, dtype: float64

In [280]:
# drop na's
df_spy.dropna(inplace = True)

In [281]:
# find return over the sample
df_spy['Return'].mean()

0.0004385772725478149

In [282]:
# specify month and year for each entry
df_spy['Year'], df_spy['Month'] = df_spy.index.year, df_spy.index.month

In [283]:
# group by year, month and find mean return of the first day
df_spy.groupby(['Year', 'Month']).apply(lambda x: x.iloc[0]).mean()

Open         1.582197e+02
High         1.592608e+02
Low          1.570921e+02
Close        1.583282e+02
Adj Close    1.318684e+02
Volume       8.746886e+07
Return       2.441323e-03
Year         2.007336e+03
Month        6.466102e+00
dtype: float64

In [284]:
# group by year, month and find mean return of the third-to-the-last trading day of the month
df_grouped = df_spy.groupby(['Year', 'Month']).apply(lambda x: x.iloc[-3]).mean()
df_grouped, df_grouped['Return']

(Open         1.589775e+02
 High         1.599253e+02
 Low          1.578749e+02
 Close        1.589248e+02
 Adj Close    1.325908e+02
 Volume       8.047841e+07
 Return       6.807369e-04
 Year         2.007336e+03
 Month        6.466102e+00
 dtype: float64,
 0.0006807369336083447)

Question 2
Create a new dataframe that includes all dates on which there is price data available for SPY.  Add the T-Bill yield (the Close column of ^IRX) to this dataframe.  Note that ^IRX is missing on a number of dates that are in SPY, such as 2016-10-10.  You can fill in these missing values using the following command:
spy['Tbill yield filled'] = spy['Tbill yield'].fillna(method='ffill')
You should take a look at how this works and think about why it is probably ok.  
Compute T-bill returns following the instructions at the bottom of this sheet.  What are the mean and standard deviation of excess SPY returns, defined as the SPY return minus the T-bill return?

In [285]:
# load datasets
irx = pd.read_csv('^IRX.csv',  parse_dates=[0], index_col=[0])
spy = pd.read_csv('SPY.csv', parse_dates=[0], index_col=[0])
irx.info(), spy.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 19450 entries, 1960-01-04 to 2022-07-26
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Open       15651 non-null  float64
 1   High       15651 non-null  float64
 2   Low        15651 non-null  float64
 3   Close      15651 non-null  float64
 4   Adj Close  15651 non-null  float64
 5   Volume     15651 non-null  float64
dtypes: float64(6)
memory usage: 1.0 MB
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 7426 entries, 1993-01-29 to 2022-07-26
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Open       7426 non-null   float64
 1   High       7426 non-null   float64
 2   Low        7426 non-null   float64
 3   Close      7426 non-null   float64
 4   Adj Close  7426 non-null   float64
 5   Volume     7426 non-null   int64  
dtypes: float64(5), int64(1)
memory usage: 406.1 KB


(None, None)

In [286]:
# add t-bill yield
spy['Tbill yield'] = irx['Close']
spy

Open        High         Low       Close   Adj Close  \
Date                                                                     
1993-01-29   43.968750   43.968750   43.750000   43.937500   25.438095   
1993-02-01   43.968750   44.250000   43.968750   44.250000   25.619024   
1993-02-02   44.218750   44.375000   44.125000   44.343750   25.673296   
1993-02-03   44.406250   44.843750   44.375000   44.812500   25.944683   
1993-02-04   44.968750   45.093750   44.468750   45.000000   26.053244   
...                ...         ...         ...         ...         ...   
2022-07-20  392.470001  396.260010  391.029999  394.769989  394.769989   
2022-07-21  394.160004  398.839996  391.630005  398.790009  398.790009   
2022-07-22  398.920013  400.179993  392.750000  395.089996  395.089996   
2022-07-25  395.750000  396.470001  393.209991  395.570007  395.570007   
2022-07-26  393.839996  394.059998  389.950012  390.890015  390.890015   

              Volume  Tbill yield  
Date                               
1993-01-29   1003200        2.900  
1993-02-01    480500        2.900  
1993-02-02    201300        2.960  
1993-02-03    529400        2.930  
1993-02-04    531500        2.900  
...              ...          ...  
2022-07-20  71843800        2.378  
2022-07-21  64903900        2.360  
2022-07-22  72130300        2.355  
2022-07-25  53631500        2.393  
2022-07-26  52642500        2.475  

[7426 rows x 7 columns]

In [287]:
# fill na's
spy['Tbill yield filled'] = spy['Tbill yield'].fillna(method='ffill')

In [288]:
# looks ok
spy['Tbill yield filled'].loc[spy.index=='2016-10-10']

Date
2016-10-10    0.31
Name: Tbill yield filled, dtype: float64

In [289]:
# compute Nt by checking difference between dates
spy['Nt'] = spy.index.to_series().diff().dt.days.fillna(1)
print(spy['Nt'])

Date
1993-01-29    1.0
1993-02-01    3.0
1993-02-02    1.0
1993-02-03    1.0
1993-02-04    1.0
             ... 
2022-07-20    1.0
2022-07-21    1.0
2022-07-22    1.0
2022-07-25    3.0
2022-07-26    1.0
Name: Nt, Length: 7426, dtype: float64


In [290]:
# compute risk-free return
spy['rf'] = spy['Tbill yield filled'].shift(1) * spy['Nt']/36500
spy['rf']

Date
1993-01-29         NaN
1993-02-01    0.000238
1993-02-02    0.000079
1993-02-03    0.000081
1993-02-04    0.000080
                ...   
2022-07-20    0.000067
2022-07-21    0.000065
2022-07-22    0.000065
2022-07-25    0.000194
2022-07-26    0.000066
Name: rf, Length: 7426, dtype: float64

In [291]:
# find spy returns
spy['Return'] = spy['Adj Close'].pct_change()
spy['Return']

Date
1993-01-29         NaN
1993-02-01    0.007113
1993-02-02    0.002118
1993-02-03    0.010571
1993-02-04    0.004184
                ...   
2022-07-20    0.006373
2022-07-21    0.010183
2022-07-22   -0.009278
2022-07-25    0.001215
2022-07-26   -0.011831
Name: Return, Length: 7426, dtype: float64

In [292]:
# find excess returns mean and SD
spy['Excess Returns'] = spy['Return'] - spy['rf']
spy.dropna(inplace=True)
spy['Excess Returns'].describe()

count    7403.000000
mean        0.000351
std         0.011886
min        -0.109444
25%        -0.004468
50%         0.000604
75%         0.005860
max         0.145180
Name: Excess Returns, dtype: float64

Question 3
What is the maximum drawdown of OAKMX?  (Hint: use adjusted closing prices for portfolio values and then the cummax() method to get the prior highs.)

In [293]:
# load dataset
oakmx = pd.read_csv('OAKMX.csv',  parse_dates=[0], index_col=[0])

In [294]:
# the cummax() method to get the prior highs
oakmx['Prior Max'] = oakmx['Adj Close'].cummax()

In [295]:
# calculate drawdowns
oakmx['Drawdowns'] = (oakmx['Adj Close'] - oakmx['Prior Max'])/oakmx['Prior Max']

In [296]:
# find min value for drowdowns (largest drawdown)
oakmx['Drawdowns'].min()

-0.56194799486071

Question 4
What is the downside deviation of OAKMX?  As a benchmark return, use the T-Bill return.


In [297]:
# calculate oakmax return
oakmx['Return'] = oakmx['Adj Close'].pct_change()
oakmx['Return'].dropna(inplace=True)

In [298]:
# find excces return for oakmx
oakmx['Excess Return'] = oakmx['Return'] - spy['rf']
stats = oakmx['Excess Return'].describe()
stats

count    7403.000000
mean        0.000364
std         0.011548
min        -0.131806
25%        -0.004426
50%         0.000670
75%         0.005460
max         0.107211
Name: Excess Return, dtype: float64

In [299]:
# find downside deviation
oakmx['min squared'] = 0
oakmx['min squared'].loc[oakmx['Excess Return']<0] = oakmx['Excess Return']**2
min2stats = oakmx['min squared'].describe()
stats.loc['Downside Deviation'] = np.sqrt(min2stats.loc['mean'])
print(stats)

count                 7403.000000
mean                     0.000364
std                      0.011548
min                     -0.131806
25%                     -0.004426
50%                      0.000670
75%                      0.005460
max                      0.107211
Downside Deviation       0.008011
Name: Excess Return, dtype: float64


C:\Users\Daniil\AppData\Local\Temp\ipykernel_20088\1687295211.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  oakmx['min squared'].loc[oakmx['Excess Return']<0] = oakmx['Excess Return']**2


Question 5
What is the average excess return on OAKMX when the previous Adjusted Closing price is above its 200-day moving average?  What about when it is below the moving average?  (Hint: use the rolling() method to get the moving average.)  Are these means statistically significant?


In [300]:
# find moving average 200 for oakmx
oakmx['MA_200'] = oakmx['Adj Close'].rolling(window=200).mean()
oakmx.dropna(inplace=True)

In [301]:
# find cases when returns are greate
oakmx['Prev_MA_200'] = oakmx['MA_200'].shift(1)
oakmx['Prev_Adj Close'] = oakmx['Adj Close'].shift(1)
above_ma = oakmx[oakmx['Prev_Adj Close'] > oakmx['Prev_MA_200']]['Excess Return']
below_ma = oakmx[oakmx['Prev_Adj Close'] <= oakmx['Prev_MA_200']]['Excess Return']


In [302]:
# What is the average excess return on OAKMX when the previous Adjusted Closing price is above its 200-day moving average?  What about when it is below the moving average?  
above_ma.mean(), below_ma.mean()

(0.0003938579981215834, 0.00026732878718344435)

In [303]:
def calc_t_stat(df):
    sample_mean = df.mean()
    sample_std = df.std()
    n = len(df)
    t_stat = sample_mean / (sample_std / n ** (1/2))
    return t_stat

calc_t_stat(above_ma), calc_t_stat(below_ma)

(3.649496029408021, 0.5901956107225281)

At significance level alpha = 0.05, t crit = 1.96, we may conclude that above_ma mean excess return is statistically signficant (reject H0, 3.64>1.96), below_ma is not significant (accept H0 0.59<1.96)

Question 6
From finance.yahoo.com, download the monthly historical financial data for ticker ARKK, which is the ARK Innovation ETF. Save this as a csv file and read it into your program. Construct a dataframe which includes the monthly returns for this ETF over the months for June 2018-June 2023. This should include a sample of 60 monthly returns. Add to this dataframe the Fama-French 3-factor data from the file ‘FFdata.csv’.  


In [304]:
import yfinance

In [305]:
# import df
arkk = pd.read_csv('ARKK.csv',  parse_dates=[0], index_col=[0])

In [306]:
# show head
arkk.head()

Open       High        Low      Close  Adj Close   Volume
Date                                                                      
2017-10-01  34.439999  35.639999  33.570000  35.509998  32.761589  2674400
2017-11-01  35.759998  38.200001  33.980000  37.200001  34.320793  4156900
2017-12-01  37.250000  39.956001  36.060001  37.080002  34.210083  7049400
2018-01-01  37.340000  42.549000  37.247002  41.240002  38.539669  8727800
2018-02-01  40.689999  42.540001  36.360001  40.910000  38.231277  8614400

In [307]:
# filter dates for arkk dataset
start = '2018-06-01'
end = '2023-06-01'
arkk = arkk[(arkk.index >= start) & (arkk.index <= end)]

In [308]:
# setup arkk to find returns
arkk.rename(columns={'Adj Close': 'rets'}, inplace=True)
arkk = arkk['rets']

In [309]:
# find returns as a series
arkk_rets = arkk.pct_change()
arkk_rets.info()

<class 'pandas.core.series.Series'>
DatetimeIndex: 61 entries, 2018-06-01 to 2023-06-01
Series name: rets
Non-Null Count  Dtype  
--------------  -----  
60 non-null     float64
dtypes: float64(1)
memory usage: 976.0 bytes


In [310]:
# upload ff dataset
ff = pd.read_csv('FFdata.csv',  parse_dates=[0])

C:\Users\Daniil\AppData\Local\Temp\ipykernel_20088\1363168765.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  ff = pd.read_csv('FFdata.csv',  parse_dates=[0])


In [311]:
# filter ff dataset
start = '201806'
end = '202306'
ff = ff[(ff['Date'] >= start) & (ff['Date'] <= end)]

In [312]:
# remake date column set as an index
import datetime
format_string = '%Y%m'
for i in ff.index:
    ff.loc[i,'datet'] = datetime.datetime.strptime(ff.loc[i,'Date'], format_string)
ff.set_index('datet',inplace=True)
ff = ff.drop(columns=['Date'])
ff

Mkt-RF   SMB   HML    RF
datet                               
2018-06-01    0.48  1.14 -2.34  0.14
2018-07-01    3.19 -2.22  0.47  0.16
2018-08-01    3.44  1.14 -3.94  0.16
2018-09-01    0.06 -2.27 -1.70  0.15
2018-10-01   -7.68 -4.78  3.40  0.19
...            ...   ...   ...   ...
2023-02-01   -2.58  1.17 -0.81  0.34
2023-03-01    2.51 -5.51 -8.85  0.36
2023-04-01    0.61 -3.35 -0.04  0.35
2023-05-01    0.35  1.61 -7.72  0.36
2023-06-01    6.46  1.54 -0.26  0.40

[61 rows x 4 columns]

In [313]:
# scale ff dataset
ff['Mkt-RF']=ff['Mkt-RF']/100
ff['SMB']=ff['SMB']/100
ff['HML']=ff['HML']/100
ff['RF']=ff['RF']/100

In [314]:
# Add to this dataframe the Fama-French 3-factor data from the file ‘FFdata.csv’.  
merged_df = pd.merge(arkk_rets, ff, how='inner', left_index=True, right_index=True)
merged_df 

rets  Mkt-RF     SMB     HML      RF
Date                                                
2018-06-01       NaN  0.0048  0.0114 -0.0234  0.0014
2018-07-01 -0.009560  0.0319 -0.0222  0.0047  0.0016
2018-08-01  0.115600  0.0344  0.0114 -0.0394  0.0016
2018-09-01 -0.047485  0.0006 -0.0227 -0.0170  0.0015
2018-10-01 -0.100549 -0.0768 -0.0478  0.0340  0.0019
...              ...     ...     ...     ...     ...
2023-02-01 -0.007764 -0.0258  0.0117 -0.0081  0.0034
2023-03-01  0.018173  0.0251 -0.0551 -0.0885  0.0036
2023-04-01 -0.109569  0.0061 -0.0335 -0.0004  0.0035
2023-05-01  0.127227  0.0035  0.0161 -0.0772  0.0036
2023-06-01  0.090146  0.0646  0.0154 -0.0026  0.0040

[61 rows x 5 columns]

In [315]:
# find excess returns
merged_df['excess ret']=merged_df['rets'] - merged_df['RF']

In [316]:
# a.What has been the average annual return, average annual excess return, and average annual standard deviation for the ARKK ETF? 
arkk_rets_anl = merged_df['rets'].mean() * 12
exs_ret_anl = merged_df['excess ret'].mean() * 12
arkk_rets_std = merged_df['rets'].std()*np.sqrt(12)
arkk_rets_anl, exs_ret_anl, arkk_rets_std

(0.0932019792120515, 0.07874197921205153, 0.4172429848579441)

In [317]:
# b.What has been the Sharpe ratio for ARKK over this period? 
arkk_sharpe = merged_df['excess ret'].mean() / merged_df['excess ret'].std()*np.sqrt(12)
arkk_sharpe

0.18876176210187934

c.	Using just the Mkt-rf factor, compute the CAPM beta for the ARKK fund. What has been the alpha for ARKK over this period? Has it been statistically significant? 

In [318]:
# dropna from merged df
merged_df = merged_df.dropna()

In [319]:
# setup linear equation for CAPM
from sklearn.linear_model import LinearRegression
model = LinearRegression()
X = merged_df['Mkt-RF']  #select the X value(s)
x=np.array(X).reshape(-1,1) 
y=np.array(merged_df['excess ret'])
model.fit(x,y)

LinearRegression()

In [320]:
# compute the CAPM beta for the ARKK fund
beta=model.coef_
print(beta)

[1.67760937]


In [321]:
# What has been the alpha for ARKK over this period? 
alpha=model.intercept_
print(alpha)

-0.009129408034746399


In [322]:
# find r squared
r_sq = model.score(x,y)
print(r_sq)

0.6168504753819245


In [323]:
# test significance
import statsmodels.api as sm
X = sm.add_constant(X)
model = sm.OLS(y, X).fit()
p_values = model.summary2().tables[1]['P>|t|']
print(p_values)

const     3.573880e-01
Mkt-RF    1.087816e-13
Name: P>|t|, dtype: float64


At significance level alpha = 0.05, we may conclude that alpha is statistically signficant (3.573880e-01)

In [324]:
# show some stats
return_above_mkt = merged_df['excess ret'].mean() - merged_df['Mkt-RF'].mean()
print('Annualized return above benchmark: ',return_above_mkt*12)
print('Annualized alpha: ', alpha*12)

Annualized return above benchmark:  -0.03349802078794849
Annualized alpha:  -0.10955289641695679


In [325]:
# run fama french model
ffmodel=LinearRegression()
X = merged_df[['Mkt-RF','SMB','HML']]  #select the X values -- here we take three regressors
x=np.array(X).reshape(-1,3) 
y=np.array(merged_df['excess ret'])
ffmodel.fit(x,y)

LinearRegression()

In [326]:
# What has been the alpha under the 3-factor model? 
ff_alpha=ffmodel.intercept_*12
ff_alpha

-0.07228225219674655

In [327]:
# What are the factor loadings? 
ff_betas = ffmodel.coef_
ff_betas

array([ 1.44063754,  1.51653622, -0.84374106])

In [328]:
# show r squared for fama french
ffmodel.score(x,y)

0.8396166263981005

In [329]:
# Is it statistically significant? 
X = sm.add_constant(X)
model = sm.OLS(y, X).fit()
p_values = model.summary2().tables[1]['P>|t|']
print(p_values)

const     3.621086e-01
Mkt-RF    6.464675e-17
SMB       6.785435e-08
HML       9.753188e-08
Name: P>|t|, dtype: float64


At significance level alpha = 0.05, we may conclude that alpha is statistically signficant (3.621086e-01)

e.	In a markdown cell, explain why the alpha from the CAPM model is different from the alpha from the FF 3-factor model. Explain how the factor loadings from the 3-factor model regression reflect the type of stocks that ARKK tends to hold in its portfolio. 

While the CAPM alpha considers only market risk, the Fama-French 3-Factor Model alpha accounts for market risk, size risk (SMB), and value risk (HML). This difference reflects how additional risk factors beyond the market influence an asset's returns. Here the negative alpha in both models signals that ARKK fund's performance was below that of the market. 

A positive factor loading on the Market (Mkt-RF) factor (1.440638) suggests that the fund tends to perform well when the broader market does.
A positive factor loading on the Size (SMB) factor (1.516536) indicates exposure to small-cap stocks, potentially benefiting from their outperformance.
The negative factor loading on the Value (HML) factor (-0.843741) implies a preference away from value-oriented stocks.